lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemFactory.java

```diff
@@ -15,160 +15,42 @@
 
 import com.google.inject.Inject;
 import io.opentelemetry.api.OpenTelemetry;
-import io.opentelemetry.instrumentation.awssdk.v2_2.AwsSdkTelemetry;
 import io.trino.filesystem.TrinoFileSystem;
 import io.trino.filesystem.TrinoFileSystemFactory;
 import io.trino.spi.security.ConnectorIdentity;
 import jakarta.annotation.PreDestroy;
-import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
-import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
-import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
-import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
-import software.amazon.awssdk.core.client.config.ClientOverrideConfiguration;
-import software.amazon.awssdk.http.apache.ApacheHttpClient;
-import software.amazon.awssdk.http.apache.ProxyConfiguration;
-import software.amazon.awssdk.regions.Region;
 import software.amazon.awssdk.services.s3.S3Client;
-import software.amazon.awssdk.services.s3.S3ClientBuilder;
-import software.amazon.awssdk.services.sts.StsClient;
-import software.amazon.awssdk.services.sts.StsClientBuilder;
-import software.amazon.awssdk.services.sts.auth.StsAssumeRoleCredentialsProvider;
-import software.amazon.awssdk.services.sts.auth.StsWebIdentityTokenFileCredentialsProvider;
 
-import java.net.URI;
-import java.util.Optional;
-import java.util.concurrent.ExecutorService;
-
-import static io.airlift.concurrent.Threads.daemonThreadsNamed;
-import static io.trino.filesystem.s3.S3FileSystemConfig.RetryMode.getRetryStrategy;
-import static io.trino.filesystem.s3.S3FileSystemConstants.EXTRA_CREDENTIALS_ACCESS_KEY_PROPERTY;
-import static io.trino.filesystem.s3.S3FileSystemConstants.EXTRA_CREDENTIALS_SECRET_KEY_PROPERTY;
-import static io.trino.filesystem.s3.S3FileSystemConstants.EXTRA_CREDENTIALS_SESSION_TOKEN_PROPERTY;
-import static java.lang.Math.toIntExact;
-import static java.util.concurrent.Executors.newCachedThreadPool;
+import java.util.concurrent.Executor;
 
 public final class S3FileSystemFactory
         implements TrinoFileSystemFactory
 {
+    private final S3FileSystemLoader loader;
     private final S3Client client;
     private final S3Context context;
-    private final ExecutorService uploadExecutor;
+    private final Executor uploadExecutor;
 
     @Inject
     public S3FileSystemFactory(OpenTelemetry openTelemetry, S3FileSystemConfig config)
     {
-        S3ClientBuilder s3 = S3Client.builder();
```

Comment:
Would be easier to review if moving of this code to the loader was extracted to a separate commit.

Reply:
I completely agree, and tried to do that originally, but the code is restructured significantly. I didn't see a good way to do a clean refactoring.